### PR TITLE
Use the private RNG for data that is not public

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -2749,8 +2749,7 @@ static int tls_process_cke_rsa(SSL *s, PACKET *pkt, int *al)
      * generates a random premaster secret for the case that the decrypt
      * fails. See https://tools.ietf.org/html/rfc5246#section-7.4.7.1
      */
-
-    if (ssl_randbytes(s, rand_premaster_secret, sizeof(rand_premaster_secret)) <= 0)
+    if (RAND_priv_bytes(rand_premaster_secret, sizeof(rand_premaster_secret)) <= 0)
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply ssl/statem/statem_srvr.c]: The client normally sends us the premaster secret encrypted to the server's RSA key. If decrypting that fails we'll end up using this random premaster secret and continue as if decryption was successful. So it's clearly not long term, in normal cases it's not even used at all.

File: ssl/ssl_lib.c [hunk 2]
```diff
@@ -2855,9 +2855,9 @@ SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth)
     /* Setup RFC5077 ticket keys */
     if ((RAND_bytes(ret->ext.tick_key_name,
                     sizeof(ret->ext.tick_key_name)) <= 0)
-        || (RAND_bytes(ret->ext.tick_hmac_key,
+        || (RAND_priv_bytes(ret->ext.tick_hmac_key,
                        sizeof(ret->ext.tick_hmac_key)) <= 0)
-        || (RAND_bytes(ret->ext.tick_aes_key,
+        || (RAND_priv_bytes(ret->ext.tick_aes_key,
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply ssl/ssl_lib.c]: That keys is used to send an encrypted ssl ticket to the client which the client can send back in the next connection. Note that this key is in an SSL_CTX, not an SSL. As far as I know we only generate this key once per SSL_CTX, so this key might be used for a lot of connections for a  long time.

**Comment by kroeckx** [comment]
The reason to start this pull request is so that people can comment on it to see in which direction it could be going.

File: ssl/ssl_lib.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mspncp) [reply ssl/ssl_lib.c]: Yes, I noticed that it's an SSL_CTX, so ssl_randbytes() can't be used. But I was not sure about the typical lifetime of an SSL_CTX. Does a webserver typically have only one or multiple SSL_CTX contexts? Does an SSL_CTX live througout the entire lifetime of the process? Is it passed on to forked children?  etc., etc...  As I said, I know too little about TLS implementation details.

Still, even if the process lives for some months, that's still much less than the lifetime of a certificate. So I thougth `RAND_bytes()` would be fine. But that's probably a matter of taste.

**Comment by kroeckx** [comment]
I've pushed a new commit with what I think the private/public split is. There are actually cases I'm not really sure about.

**Comment by richsalz** [comment]
I think there are many ways to slice this.  Keep in mind that I don't think protecting against root, or other process that can our memory is worth trying to address.  So the question becomes, can an adversary see something about one CSPRNG that gives it information about others?  To me, that means having a CSPRNG per each SSL object, or if that is too much (memory or whatever) then for each SSL_CTX object.  And then, maybe you have two at whatever level is appropriate -- one for private key material, and one for public.  But also, maybe that is too much.

I think it makes sense to have two CSPRNG's at the global level, maybe.  What "public" key information is okay to leak?  Maybe it's all private?  Maybe for CMS, the bulk encryption key is kind of public, since it's protected via asymmetric crypto.

I think more than two levels of CSPRNG are more than we need, but don't feel strongly about that. I think the second-level ones should seed off a global one, where the personalization string includes the memory address of the second-level object.

Another way to slice this is to have one (maybe two) per thread.

I have no information about which is better.  I hope whoever works on this, or the related PR's, can justify their decisions with some facgts.

File: ssl/ssl_lib.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply ssl/ssl_lib.c]: It all depends on the application, but I think it is safe to assume that SSL_CTX is long lived and passed to children.

However, this case is interesting.  An attacker would be able to grab an arbitrary amount of information from the DRBG which pushes toward using the public DRBG yet the values are long lived (process lifetime) which leans towards the private.  It almost makes for a third kind.

We could instantiate a new DRBG for this and throw it away afterwards.  Is this overkill?

**Comment by kroeckx** [comment]
So does anybody want to comment on how to go forward with this? I'm thinking about having 2 DRBGs in the SSL object, one that has as parent the public and the other the private DRBG.

Did we have a discussion about this in some other PR/issue?

**Comment by richsalz** [comment]
There was discussion in another PR; @mspncp probably remembers where.  FWIW I think two global generators is enough.  Maybe one in the SSL object to avoid any "leakage."  Right now it feels like we're headed toward a separate generator for every single time we need random bytes :)

**Comment by mspncp** [comment]
I don't remember the location, but I remember having said that the SSL drbg is in a sense already a private instance, so two SSL instances are IMHO a little overkill. I mean, we are talking about things like session keys, not long term secrets, right?

**Comment by kroeckx** [comment]
I think a session is actually also long term. I expect that my TLS session still can't be decrypted in X years.

Anyway, I think there are 2 ways that people seem to look at this:
- public vs private/secret
- long term vs short term

If you look at something like ECDHE, you create a temporary key and sign the public key. It's at least my understanding that that temporary key is important to keep secret. In #5481 @richsalz insisted that we change from RAND_bytes() to RAND_priv_bytes() for it. But I guess some people could argue that the X functions should be done with the public RNG, and the ED with the private.

For me it's usually at least clear what is considered private/secret vs public than what is long term or short term.

Anyway, I would like to move as much as possible DRBG calls use the SSL specific DRBG, which is what #5510 does. That is currently at the point that I can set a break point at RAND_bytes() and never hit it (except in SSL_CTX_new()). The only place where I currently hit RAND_priv_bytes() is in the generation of a temporary key.

But if people think this patch makes sense we would start taking the lock on the private DRBG more, and so it might make sense to have a private DRBG we don't need to lock. That could for instance be per SSL, or per thread.

**Comment by kroeckx** [comment]
How hard would it be to automatically create two thread local DRBGs?

**Comment by mspncp** [comment]
> How hard would it be to automatically create two thread local DRBGs?

Just to be sure that I understand your question right: Are you asking for the effort to replace the `<public>` and the `<private>` DRBG by per-thread DRBG instances `<public_i>` and `<private_i>` (i=1,2,...) which are automatically created and destroyed and accessed via thread local storage?

In theory, this would be a good idea, because it would make `RAND_bytes()` and `RAND_priv_bytes()` lock-free. Modifying my ascii art picture from the documentation, this would give the following overall picture:

```
           +--------------------+
           | os entropy sources |
           +--------------------+
                    |            
                    v           +-----------------------------+
  RAND_add() ==> <master>     <-| shared DRBG (with locking)  |
                  /   \         +-----------------------------+
                 |     |              +---------------------------+
          <public_i>  <private_i>  <- | per-thread DRBG instances |
                |        |            +---------------------------+
                |        + ==> RAND_priv_bytes()
                |        |      ^
                |        |      |
                |        |     +------------------------------------+
                |        |     | used only for long-term secrets    |
                |        |     | like private keys for certificates |
                |        |     +------------------------------------+
                |        |
                |        v
                |        <ssl->priv_drbg> ==> RAND_DRBG_bytes(ssl->priv_drbg, ...)
                |                ^                         
                |                |                         
                |               +-------------------------+
                |               | private per-SSL instance |
                |               +-------------------------+
                |
                + ==> RAND_bytes()
                |      ^
                |      |
                |     +----------------------------------+
                |     | general purpose random generator |
                |     +----------------------------------+
                v
              <ssl->drbg> ==> RAND_DRBG_bytes(ssl->drbg, ...)
                     ^
                     |
                    +-------------------------+
                    | public per-SSL instance |
                    +-------------------------+
```

The drawback is that thread local storage is platform and compiler specific and might not be available everywhere. It would have to be an optional feature which could be selectively enabled on suitable platforms. So it's hard to say how big the effort is. Does OpenSSL currently use thread local storage at all?

At first sight, the above picture looks a bit over-engineered, so I'm undeceded whether it's worth the effort. If at all, this would be a task for the next release. Let's discuss this after the code freeze in more detail.

**Comment by kroeckx** [comment]
We have thread support, including thread local storage, for pthread and windows. If it's not enabled, we don't even take a lock.

**Comment by mspncp** [comment]
So is it  that what I sketched in my last post what you were thinking of?

**Comment by kroeckx** [comment]
It's at least one way of doing it.

**Comment by paulidale** [comment]
I'm okay moving the two global drbgs to be per thread.  I think I mentioned per thread drbg's at one point but I was think of them in addition to the two global ones rather than replacing them.

I don't see a necessity for two per ssl drbg's but wouldn't object to having them.  The greater the separation the better I feel -- so long as the overhead isn't huge -- which it isn't.  @richsalz I wouldn't want to see a per random request drbg either.

File: crypto/bn/bn_prime.c [hunk 3]
```diff
@@ -279,6 +279,7 @@ static int probable_prime(BIGNUM *rnd, int bits, prime_t *mods)
     char is_single_word = bits <= BN_BITS2;
 
  again:
+    /* XXX: Not all primes are private */
     if (!BN_priv_rand(rnd, bits, BN_RAND_TOP_TWO, BN_RAND_BOTTOM_ODD))
```
Comment (mspncp) [inline crypto/bn/bn_prime.c]: The comment `Not all primes are private` right in front of a call to `BN_priv_rand()` is a little confusing. And by the way: What does `XXX:` mean?

File: crypto/bn/bn_sqrt.c [hunk 4]
```diff
@@ -179,7 +179,7 @@ BIGNUM *BN_mod_sqrt(BIGNUM *in, const BIGNUM *a, const BIGNUM *p, BN_CTX *ctx)
             if (!BN_set_word(y, i))
                 goto end;
         } else {
-            if (!BN_rand(y, BN_num_bits(p), 0, 0))
+            if (!BN_priv_rand(y, BN_num_bits(p), 0, 0))
```
Comment (mspncp) [inline crypto/bn/bn_sqrt.c]: I'm not sure whether I always understand the rationale behind your choice of public vs. private rng.

File: crypto/bn/bn_prime.c [hunk 5]
```diff
@@ -363,7 +364,7 @@ int bn_probable_prime_dh(BIGNUM *rnd, int bits,
     if ((t1 = BN_CTX_get(ctx)) == NULL)
         goto err;
 
-    if (!BN_priv_rand(rnd, bits, BN_RAND_TOP_ONE, BN_RAND_BOTTOM_ODD))
+    if (!BN_rand(rnd, bits, BN_RAND_TOP_ONE, BN_RAND_BOTTOM_ODD))
```
Comment (mspncp) [inline crypto/bn/bn_prime.c]: Why do you use the public drbg in `bn_probable_prime_dh()`and the private drbg in `probable_prime()`?

File: ssl/ssl_lib.c [hunk 6]
```diff
@@ -5141,7 +5141,7 @@ uint32_t SSL_get_max_early_data(const SSL *s)
     return s->max_early_data;
 }
 
-int ssl_randbytes(SSL *s, unsigned char *rnd, size_t size)
+int ssl_randbytes_public(SSL *s, unsigned char *rnd, size_t size)
```
[the remark this answers is not in the document]
Comment (mspncp) [reply ssl/ssl_lib.c]: `ssl_randbytes()` has been obsoleted by 16cfc2c90d9e7776965db07c1f31bbec2f6c41e3. This needs to be rebased.

**Comment by mspncp** [comment]
I will continue the review after the pull request is rebased.

File: crypto/bn/bn_prime.c (hunk 3 above)
Comment (kroeckx) [reply crypto/bn/bn_prime.c]: Maybe I should replace it with TODO

File: crypto/bn/bn_sqrt.c (hunk 4 above)
Comment (kroeckx) [reply crypto/bn/bn_sqrt.c]: Anything that gets send to a peer is public, the rest is private.

File: crypto/bn/bn_prime.c (hunk 5 above)
Comment (kroeckx) [reply crypto/bn/bn_prime.c]: That doesn't make sense at all ...

Comment (kroeckx) [reply crypto/bn/bn_prime.c]: (And clearly why the other one is marked as XXX/TODO)

Comment (kroeckx) [reply crypto/bn/bn_prime.c]: Oh, but this is about primes for use in DH, where the prime is actually send to the peer.

**Comment by mspncp** [comment]
Is this still WIP or do you consider it ready-for-review?

**Comment by kroeckx** [comment]
It's WIP because I didn't go over all the calls yet.

**Pull request closed without merging**